desktop/core/src/desktop/js/utils/hooks/useFileUpload/useChunkUpload.ts

```diff
@@ -26,169 +26,194 @@ import {
-    },
-    onError
+    }
   });
 
-  const updateItemStatus = (serverResponse: TaskServerResponse[]) => {
-    const statusMap = getStatusHashMap(serverResponse);
+  const { reloadData: getAvailableSpace } = useLoadData<AvailableServerSpaceResponse>(
+    UPLOAD_AVAILABLE_SPACE_URL,
+    { skip: true } // Call will be triggered via reloadData function on need basis
+  );
 
-    const remainingItems = awaitingStatusItems.filter(item => {
-      const status = statusMap[item.uuid];
-      if (status === TaskStatus.Success || status === TaskStatus.Failure) {
-        const ItemStatus =
-          status === TaskStatus.Success ? FileUploadStatus.Uploaded : FileUploadStatus.Failed;
-        onStatusUpdate(item, ItemStatus);
-        return false;
-      }
-      return true;
-    });
-    if (remainingItems.length === 0) {
-      onComplete();
-    }
-    setAwaitingStatusItems(remainingItems);
+  const onError = (chunkItem: UploadChunkItem) => (error: Error) => {
+    onItemUpdate(chunkItem.uuid, { status: FileUploadStatus.Failed, error });
   };
 
-  const { data: tasksStatus } = useLoadData<TaskServerResponse[]>(
-    '/desktop/api2/taskserver/get_taskserver_tasks/',
-    {
-      pollInterval: awaitingStatusItems.length ? 5000 : undefined,
-      skip: !awaitingStatusItems.length,
-      transformKeys: 'none'
-    }
-  );
-
-  useEffect(() => {
-    if (tasksStatus) {
-      updateItemStatus(tasksStatus);
-    }
-  }, [tasksStatus]);
-
-  const onChunksUploadComplete = async () => {
-    if (processingItem) {
-      const { url, payload } = getChunksCompletePayload(processingItem, chunkSize);
-      return save(payload, {
-        url,
-        onSuccess: onSuccess(processingItem)
+  const processChunkUploadStatus = (tasksStatus?: TaskServerResponse[]) => {
+    if (tasksStatus?.length) {
+      const statusMap = getStatusHashMap(tasksStatus);
+      setTaskServerItems(prev => {
+        Object.keys(prev).forEach(uuid => {
+          const status = statusMap[uuid];
+          if (status === TaskStatus.Success || status === TaskStatus.Failure) {
+            const ItemStatus =
+              status === TaskStatus.Success ? FileUploadStatus.Uploaded : FileUploadStatus.Failed;
+            onItemUpdate(uuid, { status: ItemStatus });
+            delete prev[uuid];
+          }
+        });
+        if (Object.keys(prev).length === 0) {
+          onComplete();
+        }
+        return prev;
       });
     }
   };
 
-  const uploadChunk = async (chunkItem: UploadChunkItem) => {
-    const { url, payload } = getChunkItemPayload(chunkItem, chunkSize);
-    return save(payload, { url });
+  useLoadData<TaskServerResponse[]>('/desktop/api2/taskserver/get_taskserver_tasks/', {
+    pollInterval: 5000,
+    skip: Object.keys(taskServerItems).length === 0,
+    onSuccess: processChunkUploadStatus,
+    transformKeys: 'none'
+  });
+
+  const onChunksUploadComplete = async (chunkItem: UploadChunkItem) => {
```

Comment:
A more conventional approach is to call these handleChunksUploadComplete, handleChunkUploadSuccess, handleUploadProgress etc to separate them from the prop callbacks.